File: problem3/problem3.cpp
class FizzBuzz {
private:
    int n; // The maximum number to check
    int currentNumber = 1; // Current number being processed

    mutex mtx; // Mutex for thread synchronization
    condition_variable cv; // Condition variable for thread synchronization

public:
    // Constructor to initialize 'n'
    FizzBuzz(int n) {
        this->n = n;
    }

    // Function for handling "fizz" printing
    void fizz(function<void()> printFizz) {
        wrapper(printFizz, [this](){return currentNumber%3 == 0 && currentNumber%5 != 0;});
    }

    // Function for handling "buzz" printing
    void buzz(function<void()> printBuzz) {
        wrapper(printBuzz, [this](){return currentNumber%3 != 0 && currentNumber%5 == 0;});
    }

    // Function for handling "fizzbuzz" printing
    void fizzbuzz(function<void()> printFizzBuzz) {
        wrapper(printFizzBuzz, [this](){return currentNumber%3 == 0 && currentNumber%5 == 0;});
    }

    // Function for handling number printing
    void number(function<void(int)> printNumber) {
        wrapper([printNumber, this](){ return printNumber(currentNumber); },
                [this](){return currentNumber%3 != 0 && currentNumber%5 != 0;});
    }

private:
    // Helper function to run the task with proper synchronization
    void wrapper(function<void()> task, function<bool()> condition) {
        while (true) {
            {
                // Lock the mutex
                unique_lock<mutex> lock(mtx);

                // Wait until the condition is satisfied or the number exceeds 'n'
                cv.wait(lock, [condition, this]() { return (condition() || (currentNumber > n)); });

                // If the number exceeds 'n', notify all threads and exit the loop
                if (currentNumber > n) {
                    cv.notify_all();
                    return;
                }
            }

            // Execute the print function
            task();

            {
                // Lock the mutex to update the current number
                unique_lock<mutex> lock(mtx);
                currentNumber += 1;
            }

            // Notify all threads about the state update
            cv.notify_all();
        }
    }
};

File: problem3/problem3/problem3.java
public class FizzBuzz {
    private final int n; // The maximum number to check
    private int currentNumber = 1; // Current number being processed

    // Constructor to initialize `n`
    public FizzBuzz(int n) {
        this.n = n;
    }

    // Function for handling "fizz" printing
    public void fizz(Runnable printFizz) throws InterruptedException {
        wrapper(printFizz, () -> currentNumber % 3 == 0 && currentNumber % 5 != 0);
    }

    // Function for handling "buzz" printing
    public void buzz(Runnable printBuzz) throws InterruptedException {
        wrapper(printBuzz, () -> currentNumber % 3 != 0 && currentNumber % 5 == 0);
    }

    // Function for handling "fizzbuzz" printing
    public void fizzbuzz(Runnable printFizzBuzz) throws InterruptedException {
        wrapper(printFizzBuzz, () -> currentNumber % 3 == 0 && currentNumber % 5 == 0);
    }

    // Function for handling number printing
    public void number(IntConsumer printNumber) throws InterruptedException {
        wrapper(() -> printNumber.accept(currentNumber), () -> currentNumber % 3 != 0 && currentNumber % 5 != 0);
    }

    // Helper function to run the task with proper synchronization
    private void wrapper(Runnable task, Condition condition) throws InterruptedException {
        while (true) {
            // Acquire the lock to wait and check the condition
            synchronized (this) {
                while (!(condition.test() || currentNumber > n)) {
                    this.wait();
                }

                // If the number exceeds `n`, notify all threads and exit the loop
                if (currentNumber > n) {
                    notifyAll();
                    return;
                }
            }

            // Execute the print function without holding the lock
            task.run();

            // Acquire the lock again to update the current number and notify
            synchronized (this) {
                currentNumber += 1;
                // Notify all threads about the state update
                this.notifyAll();
            }
        }
    }

    // Interface for the condition check
    private interface Condition {
        boolean test();
    }
}
